app/grandchallenge/algorithms/tasks.py

```diff
@@ -14,122 +13,61 @@
-        )
-
-    if user_upload_pks:
-        assignment_tasks.extend(
-            add_file_to_component_interface_value.signature(
-                kwargs={
-                    "component_interface_value_pk": civ_pk,
-                    "user_upload_pk": user_upload_pk,
-                    "target_pk": job.algorithm_image.algorithm.pk,
-                    "target_app": "algorithms",
-                    "target_model": "algorithm",
-                },
-                immutable=True,
-            )
-            for civ_pk, user_upload_pk in user_upload_pks.items()
-        )
-
-    canvas = chain(
-        group(assignment_tasks),
-        execute_algorithm_job_for_inputs.signature(
-            kwargs={"job_pk": job_pk}, immutable=True
-        ),
-    )
-
-    on_commit(canvas.apply_async)
-
-
-@acks_late_micro_short_task
+@acks_late_micro_short_task(retry_on=(LockNotAcquiredException,))
 @transaction.atomic
 def execute_algorithm_job_for_inputs(*, job_pk):
-    from grandchallenge.algorithms.models import Job
-
-    job = Job.objects.get(pk=job_pk)
+    job = lock_model_instance(
+        app_label="algorithms", model_name="job", pk=job_pk
+    )
 
     # Notify the job creator on failure
     linked_task = send_failed_job_notification.signature(
         kwargs={"job_pk": str(job.pk)}, immutable=True
     )
 
-    # check if all ComponentInterfaceValue's have a value.
-    missing_inputs = list(civ for civ in job.inputs.all() if not civ.has_value)
-
-    if missing_inputs:
-        job.update_status(
-            status=job.CANCELLED,
-            error_message=(
-                f"Job can't be started, input is missing for "
-                f"{oxford_comma([c.interface.title for c in missing_inputs])}"
-            ),
-        )
-        on_commit(linked_task.apply_async)
-    else:
+    if job.inputs_complete and job.status == job.PENDING:
         job.task_on_success = linked_task
+        job.status = job.INPUTS_VALIDATED
         job.save()
         on_commit(
             execute_algorithm_job.signature(
                 kwargs={"job_pk": job_pk}, immutable=True
             ).apply_async
         )
+    else:
+        logger.info(
+            "Nothing to do: the inputs are still being created and validated, "
+            "or the job is already being executed."
+        )
+        return
 
 
 @acks_late_micro_short_task(retry_on=(TooManyJobsScheduled,))
 @transaction.atomic
 def execute_algorithm_job(*, job_pk):
     from grandchallenge.algorithms.models import Job
 
-    if Job.objects.active().count() >= settings.ALGORITHMS_MAX_ACTIVE_JOBS:
+    job = Job.objects.get(pk=job_pk)
+    if not job.status == Job.INPUTS_VALIDATED:
+        raise RuntimeError("Job is not ready for execution")
+    elif Job.objects.active().count() >= settings.ALGORITHMS_MAX_ACTIVE_JOBS:
         raise TooManyJobsScheduled
     else:
-        job = Job.objects.get(pk=job_pk)
+        job.status = Job.PENDING
+        job.save()
```

Comment:
Ok, I followed the pattern we use for evaluations, where we also re-use the PENDING state after executing algorithms is done, but I get your point that that state should be unique. What you suggest makes sense and means that a job would be in "validating inputs" state from creation onwards, unless we add yet another new state to distinguish pending from pending for execution. In other words, I see two ways of state progression after job creation:

1) VALIDATING_INPUTS -> PENDING -> PROVISIONING -> PROVISIONED etc. 
2) PENDING -> VALIDATING_INPUTS -> PENDING_FOR_EXECUTION -> PROVISIONING -> PROVISIONED etc. 

I think option 1 is good enough since validating inputs starts right away (i.e "create_civ" is called synchronously as soon as we create the job, so there is little point in having an initial "PENDING" state). I will proceed with implementing that

Reply:
> Is raising a RuntimeError correct here? We're definitely only expecting this job to be called once?

This task should only be called once, when inputs are complete. We make sure it is only scheduled once by checking that the status is `job.VALIDATING_INPUTS` before changing it to `INPUTS_VALIDATED` and calling this function in `execute_algorithm_job_for_inputs.` A job can only go from validating inputs to inputs validated once and so this task will only be called once, or else we raise an error. Does that make sense? 

Manually requeuing a job in the admin will not call this method, but will directly call the execute method on the job.

Reply:
Having looked at this again, I actually don't see the need for this separate `execute_algorithm_job` task, it can all be done in `execute_algorithm_job_for_inputs` I think. Then we also don't need the `INPUTS_VALIDATED` status, and can make do with just `VALIDATING_INPUTS` similar to how we handle evaluations. A job now goes from `VALIDATING_INPUTS` to `PENDING` to `PROVISIONING` etc.